src/core/api.rs

```diff
@@ -40,13 +40,299 @@ impl PartialEq for Covercrypt {
+            })
+            .transpose()?;
+        Ok((
+            SymmetricKey::try_from_slice(&seed)?,
+            Self {
+                encapsulation,
+                encrypted_metadata,
+                phantom: PhantomData,
+            },
+        ))
+    }
+
+    /// Decrypts the header with the given user secret key.
+    ///
+    /// The nonce used is extracted from the encapsulation tag.
+    ///
+    /// - `cover_crypt`         : `Covercrypt` object
+    /// - `usk`                 : `Covercrypt` user secret key
+    /// - `authentication_data` : authentication data used in the DEM encryption
+    fn decrypt(
+        &self,
+        cover_crypt: &Covercrypt,
+        usk: &UserSecretKey,
+        authentication_data: Option<&[u8]>,
+    ) -> Result<Option<CleartextHeader>, Error> {
+        if let Some(seed) = CovercryptKEM::decaps(cover_crypt, usk, self.encapsulation.clone()) {
+            let metadata = self
+                .encrypted_metadata
+                .as_ref()
+                .map(|ciphertext| {
+                    let mut sym_key = Secret::<KEY_LENGTH>::default();
+                    kdf256!(&mut sym_key, &seed);
+                    E::decrypt(&sym_key, authentication_data, ciphertext)
+                })
+                .transpose()?;
+            Ok(Some(CleartextHeader {
+                symmetric_key: SymmetricKey::try_from_slice(&seed)?,
+                metadata,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// Structure containing all data encrypted in an `EncryptedHeader`.
+///
+/// - `symmetric_key`   : DEM key
+/// - `metadata`        : additional data symmetrically encrypted in a header
+#[derive(Debug, PartialEq, Eq)]
+pub struct CleartextHeader {
+    pub symmetric_key: SymmetricKey<SEED_LENGTH>,
+    pub metadata: Option<Vec<u8>>,
+}
+
+pub trait CovercryptKEM {
     /// Sets up the Covercrypt scheme.
     ///
-    /// Generates a MSK and a MPK with a tracing level of [`MIN_TRACING_LEVEL`](core::MIN_TRACING_LEVEL).
+    /// Generates a MSK and a MPK with a tracing level of
+    /// [`MIN_TRACING_LEVEL`](core::MIN_TRACING_LEVEL).
     /// They only hold keys for the origin coordinate: only broadcast
     /// encapsulations can be created.
-    pub fn setup(&self) -> Result<(MasterSecretKey, MasterPublicKey), Error> {
+    fn setup(&self) -> Result<(MasterSecretKey, MasterPublicKey), Error>;
+
+    /// Updates the MSK according to this policy. Returns the new version of the
+    /// MPK.
+    ///
+    /// Sets the MPK coordinates to the one defined by the policy:
+    /// - removes coordinates from the MSK that don't belong to the new policy
+    /// along with their associated keys;
+    /// - adds the policy coordinates that don't belong yet to the MSK,
+    /// generating new keys.
+    ///
+    /// The new MPK holds the latest public keys of each coordinates of the new policy.
+    fn update_master_keys(
```

Comment:
key management should not be part of the KEM interface.

Reply:
Similarly: should it be in PKE?